### PR TITLE
Set up conda package

### LINKED ISSUE #40: Create replace_landsurface conda package
Package the repo into a conda package to be installed in the [access-ram conda environment](https://github.com/ACCESS-NRI/access-ram-condaenv).

### PR DESCRIPTION
**Pull request by atteggiani**
Fixes #40 

- [x] Added conda build and dev environments
- [x] Added conda recipe for conda package
- [x] Added pyproject.toml for pip installation in dev mode 
- [x] Added versioneer files for package versioning 
- [x] Added conda build verify to CI
- [x] Set up CD for conda package upload to accessnri anaconda.org.

I haven't touched the main scripts in this PR, and the new package has the `hres_ic` and `hres_eccb` entry points, to be called from the rose suite (to be changed in the `nci_hres_ic` and `nci_hres_eccb` tasks `rose-app.conf` files, respectively).
In the future, it might be good to change how the scripts are run, to only have one entry point (for more information see #31).

### CODE REVIEW

**Review by engelca — CHANGES_REQUESTED** [review]
Hi @atteggiani,

To verify the changes can you/we please test the new conda environment on the supercomputer?

I will be happy to approve the changes once I compare the output with that from before the conda package.

Let me know if I can help.

**Comment by atteggiani** [comment]
> Hi @atteggiani,
> 
> To verify the changes can you/we please test the new conda environment on the supercomputer?

Yes, I agree we should test the suite with a dev `conda` environment where we installed a local build of the `conda` packages (`replace_landsurface` and `era5grib_parallel`), and verify that the suite runs fine. 

The current issue is that, on _Gadi_, I haven't been able to use `conda build` together with `versioneer` (the package to retrieve versions) because this package should be installed in the root (base) `conda` environment, but on _Gadi_ that is not the case and we cannot change it. 
I will have to check what is the solution there.

> I will be happy to approve the changes once I compare the output with that from before the conda package.

I am happy for you to test that the outputs before and after the `conda` package match. 
However, I don't think this would be strictly necessary (as long as we make sure that the suite runs), because there are no functional changes in the python scripts.

**Comment by atteggiani** [comment]
Also to note, the current CI fails because:
- CI / pre-commit: fails because of linting/formatting problems in the scripts that should be addressed, but not in this PR;
- CI / Conda Build (pull_request) fails during test of the entry points, because the `ROSE_DATA` environment variable only exists within the suite run, therefore we get [this error](https://github.com/ACCESS-NRI/replace_landsurface/actions/runs/11336803999/job/31527342545?pr=42#step:5:413). 
The build still works. 
Also, the error will be solve after we refactor the scripts as mentioned in #41.

**Comment by engelca** [comment]
@atteggiani, I think that @CodeGat may be a better reviewer for this change given the focus on the conda environment.  Would that be alright?

**Review by aidanheerdegen — CHANGES_REQUESTED** [review]
Looks good, thanks. A few requested changes, nothing major.

File: .github/workflows/CD.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        with:
+            fetch-tags: true
+            fetch-depth: 0 
+
+      - name: Setup conda environment
+        uses: conda-incubator/setup-miniconda@11b562958363ec5770fef326fe8ef0366f8cbf8a
+        with:
+          miniconda-version: "latest"
+          python-version: ${{ env.PY_VERSION }}
+          environment-file: conda/env_build.yml
```
Comment (CodeGat) [inline .github/workflows/CD.yml:28]: Should this be:
```suggestion
          environment-file: .conda/env_build.yml
```

File: .github/workflows/CD.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+          auto-update-conda: false
+          auto-activate-base: false
+          show-channel-urls: true
+
+      - name: Build and upload the conda package
+        uses: uibcdf/action-build-and-upload-conda-packages@c6e7a90ad5e599d6cde76e130db4ee52ad733ecf
+        with:
+          meta_yaml_dir: conda
```
Comment (CodeGat) [inline .github/workflows/CD.yml:36]:
```suggestion
          meta_yaml_dir: .conda
```
Similarly, should this be pointing to the `.conda` folder?

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        shell: bash -el {0}
+        run: conda-verify .conda --ignore C2105,C2122
+
+      - name: Run conda build
+        shell: bash -el {0}
+        run: conda build . --no-anaconda-upload --output-folder=./build -c conda-forge -c accessnri -c coecms
+
+      - name: Verify conda package
+        shell: bash -el {0}
+        run: conda-verify ./build/noarch/*.tar.bz2 --ignore C1105,C1115,C1141
```
Comment (CodeGat) [inline .github/workflows/CI.yml:72]: It would be good to have a comment on the why/what we're ignoring here

File: .github/workflows/CI.yml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+          miniconda-version: "latest"
+          python-version: ${{ env.PY_VERSION }} 
+          environment-file: .conda/env_build.yml
+          auto-activate-base: false
+          auto-update-conda: false
+          show-channel-urls: true
+
+      - name: Verify conda recipe
+        shell: bash -el {0}
+        run: conda-verify .conda --ignore C2105,C2122
```
Comment (CodeGat) [inline .github/workflows/CI.yml:64]: Similarly here, it would be good to have a comment on why we're ignoring these errors.

**Review by CodeGat — CHANGES_REQUESTED** [review]
Just requesting changes on the `.conda` vs `conda` stuff.

**Comment by aidanheerdegen** [comment]
> * **CI / Conda Build (pull_request)** fails during test of the entry points, because the `ROSE_DATA` environment variable only exists within the suite run, therefore we get [this error](https://github.com/ACCESS-NRI/replace_landsurface/actions/runs/11336803999/job/31527342545?pr=42#step:5:413).
>   The build still works.
>   Also, the error will be solve after we refactor the scripts as mentioned in [Change `BARRA_DIR`/`ERA_DIR` to `source_dir` and take it as a command-line argument #41](https://github.com/ACCESS-NRI/replace_landsurface/issues/41).

It's bad practice to have CI tests that you know are failing and have to override to merge. It leads to errors where things get merged that shouldn't because you get used to overriding errors. Ask me how I know. And it happened again recently.

So we should either specify a `ROSE_DATA` dummy environment variable, or change the code to give this a default value:
```python
ROSE_DATA = os.environ.get('ROSE_DATA', '')
```
or turn off the entry point tests (least desirable option)

**Comment by atteggiani** [comment]
> It's bad practice to have CI tests that you know are failing and have to override to merge. It leads to errors where things get merged that shouldn't because you get used to overriding errors. Ask me how I know. And it happened again recently.

I agree, 
for the moment setting it to a "dummy" blank variable would work in my opinion, and still preserve the functionality on _Gadi_ so I would go with that option for now.

File: .conda/env_dev.yml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+channels:
+    - conda-forge
+    - accessnri
+    - coecms
+    - default
+    - nodefaults
+
+dependencies:
+    - python >=3.10,<=3.12
```
[the remark this answers is not in the document]
Comment (atteggiani) [reply .conda/env_dev.yml:11]: Mainly not to restrict too much the installation of the package. 
If any specified version of Python is supported by the package (and currently it seems to be, even though testing might be needed to be 100% sure about it) I would leave this less restrictive.

**Review by CodeGat — APPROVED** [review]
Looks good to me now. Thanks.

**Review by CodeGat — DISMISSED** [review]
If the `pre-commit` failure isn't going to be addressed in this PR (because it would be a massive amount of lines changed or something), it should be the next cab off the rank, in my opinion. 

I'd be happy to review that, when it comes.

File: .conda/env_dev.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (aidanheerdegen) [reply .conda/env_dev.yml:9]: From a release point of view it's usually better to be more definite about what is being installed. And changes should be deliberate. 

From an organisational point of view I think we should agree on foundational things like the versions of `python` we're deploying so everyone knows what is supported. It is not uncommon for new features to be brought into the language, and it's important that they aren't used without being widely supported in the organisation.

File: .vscode/settings.json
Comment (aidanheerdegen) [inline .vscode/settings.json:1]: We shouldn't be committing `.vscode` settings.

**Review by aidanheerdegen — CHANGES_REQUESTED** [review]
Should remove that `.vscode` settings file I think.

File: .conda/env_dev.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (aidanheerdegen) [reply .conda/env_dev.yml:9]: Oh sorry, I just realised this is the dev container. What version are we specifying for actual deployment?

[the remark this answers is not in the document]
Comment (atteggiani) [reply .conda/env_dev.yml:9]: Oh I just realised this was for the development environment and not for the `conda` package, sorry.
My previous comment doesn't make much sense then. 
I agree it's better to converge to a single version of `python` for the deployed environments. 

The problem is that currently there are 3 repos connected to each other: 
- [replace_landsurface](https://github.com/ACCESS-NRI/replace_landsurface) (this repo)
- [era5_grib_parallel](https://github.com/ACCESS-NRI/era5_grib_parallel)
- [access-ram-condaenv](https://github.com/ACCESS-NRI/access-ram-condaenv)

Currently, `access-ram-condaenv` has its own [env.yml](https://github.com/ACCESS-NRI/access-ram-condaenv/blob/main/env.yml) and [env-dev.yml](https://github.com/ACCESS-NRI/access-ram-condaenv/blob/main/env-dev.yml) that are the environments that are actually used for the deployment _on Gadi_. These are the environments that need to be up to date with changes in `replace_landsurface` and `era5grib_parallel`, and should satisfy both of these (future) packages dependencies.

Therefore, I propose to:
- get rid of the `env_dev.yml` in this repo and in [era5_grib_parallel](https://github.com/ACCESS-NRI/era5_grib_parallel/blob/davide/conda_package/.conda/env_dev.yml).
- write instructions for development of both `replace_landsurface` and `era5grip_parallel` packages to use the `env_dev.yaml` in the `access-ram-condaenv` repo as a base conda environment to install the respective development package (via `pip install -e`)
- Find a good way to keep the environment files in `access-ram-condaenv` up to date with the changes in both `replace_landsurface` and `era5grip_parallel` packages (i.e. repos) (I am still not sure about the best approach for this one.)

What do you think?

[the remark this answers is not in the document]
Comment (atteggiani) [reply .conda/env_dev.yml:9]: And also set specific dependencies versions in the `access-ram-condaenv` environment files.

[the remark this answers is not in the document]
Comment (aidanheerdegen) [reply .conda/env_dev.yml]: Sounds like a good plan. Sorry if this response was a bit late.

**Comment by atteggiani** [comment]
Closed as this PR has been substituted by #45.

**Pull request closed without merging**